app/Http/Controllers/Service/Admin/VouchersController.php

```diff
@@ -47,20 +50,63 @@ public function create()
      */
     public function storeBatch(Request $request)
     {
-        $new_vouchers = [];
+        $newVouchers = [];
+
+        $sponsorIds = \App\Sponsor::all()->pluck('id')->toArray();
+
+        $voucherRules = [
+            'sponsor_id' => [
+                'required',
+                Rule::in($sponsorIds)
+            ],
+            'start' => 'required|integer|between:1,99999999',
+            'end' => [
+                'required',
+                'integer',
+                'between:1,99999999',
+                'ge_field:start'
+            ],
+        ];
+
+        $messages = [
+            'in' => 'The :atrribute must be a valid selection',
+            'ge_field' => 'The :attribute must be greater than or equal to :field'
+        ];
+
+        Validator::make($request->all(), $voucherRules , $messages)->validate();
+
         $codes = range($request['start'], $request['end']);
+
         $sponsor = Sponsor::find($request['sponsor_id'])->first();
+        $shortcode = $sponsor->shortcode;
+        $nowTime = Carbon::now();
         foreach ($codes as $c) {
             $v = new Voucher();
-            $v->code = $sponsor->shortcode . $c;
+            $v->code = $shortcode . $c;
             $v->sponsor_id = $request['sponsor_id'];
             $v->currentstate = 'requested';
-            $v->created_at = Carbon::now();
-            $v->updated_at = Carbon::now();
-            $new_vouchers[] = $v->attributesToArray();
+            $v->created_at = $nowTime;
+            $v->updated_at = $nowTime;
+            $newVouchers[] = $v->attributesToArray();
+        }
+        // batch insert.
+        // Todo : there's NO "this voucher already exists" checking!!
+        Voucher::insert($newVouchers);
```

Comment:
Would `$vouchers = Voucher::insert($newVouchers);` get them? Or does that return some db object? Or a boolean?